### PR TITLE
Add dockerfile for easier local deployment

### LINKED ISSUE #11: Add `Dockerfile`
@SantiagoTorres suggested that we dockerize the layout web tool, so that project owner can easily host their own instance, e.g. if they don't want to submit information about their software supply chain to the instance that we are hosting.

### PR DESCRIPTION
**Pull request by ShikherVerma**
Fixes #11 

Build image from docker file using
```
docker build -t="python-flask" .
```

Run
```
docker run --net=host -d python-flask
```

Notes : On arch linux I had to run it with `--net=host` for port mapping to work. Might not be needed for other distros.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM ubuntu:latest
```
Comment (SantiagoTorres) [inline Dockerfile]: I'd presonally use `debian:stable-slim`. Any reason as to why we use ubuntu in this case?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+MAINTAINER Shikher Verma "root@shikherverma.com"
```
Comment (SantiagoTorres) [inline Dockerfile]: `MAINTAINER` is deprecated, please use `LABEL`

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+RUN apt-get update -y --fix-missing
```
Comment (SantiagoTorres) [inline Dockerfile]: We may want to merge all of the layers that you have down there.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+RUN apt-get install -y python python-pip python-dev build-essential git nodejs npm ruby ruby-dev
```
Comment (SantiagoTorres) [inline Dockerfile]: You seem to use nodejs to build/install gem stuff. We probably could use a mullti-stage build in this case.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+# Install Dependencies
+RUN \
+  apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 7F0CEB10 && \
+  echo 'deb http://downloads-distro.mongodb.org/repo/ubuntu-upstart dist 10gen' > /etc/apt/sources.list.d/mongodb.list && \
+  apt-get update && \
+  apt-get install -y mongodb-org && \
+  rm -rf /var/lib/apt/lists/*
```
Comment (SantiagoTorres) [inline Dockerfile]: This kinda trickery is not needed anymore, as you can use multi-stage builds.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
+FROM ubuntu:latest
+MAINTAINER Shikher Verma "root@shikherverma.com"
+RUN apt-get update -y --fix-missing
+RUN apt-get install -y python python-pip python-dev build-essential git nodejs npm ruby ruby-dev
+# Install Dependencies
+RUN \
+  apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 7F0CEB10 && \
```
Comment (SantiagoTorres) [inline Dockerfile]: You probably want to use hkps? Does the ubuntu keyserver not support it?

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  echo 'deb http://downloads-distro.mongodb.org/repo/ubuntu-upstart dist 10gen' > /etc/apt/sources.list.d/mongodb.list && \
+  apt-get update && \
+  apt-get install -y mongodb-org && \
```
Comment (SantiagoTorres) [inline Dockerfile]: I'd suggest to add --no-install-recommends to keep the final image minimal (and lower the build times). Using DEBIAN_FRONTEND=noninteractive also avoids getting stuck on configuration steps that require choices (which usually happens when you need to setup a password for a database or so).

File: docker-start.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+mkdir -p /data/db
+mongod &
+mkdir instance
+echo 'DEBUG=False' >> instance/config.py
+echo "SECRET_KEY='?\xbf,\xb4\x8d\xa3<\x9c\xb0@\x0f5\xab,w\xee\x8d$0\x13\x8b83'" >> instance/config.py
```
Comment (SantiagoTorres) [inline docker-start.sh]: I'm not exactly fond of setting "a" secret key there, but I guess it's ok for the sake of development.

Should we aim to have this docker image eventually be good for deployment though?

File: docker-start.sh [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+mkdir -p /data/db
```
Comment (SantiagoTorres) [inline docker-start.sh]: I wonder if this folder can be made during the docker build.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  rm -rf /var/lib/apt/lists/*
+COPY . /app
+WORKDIR /app
+RUN pip install -r requirements.txt
+RUN npm install --verbose
+RUN gem install sass
```
Comment (SantiagoTorres) [inline Dockerfile]: We probably want to chain these with an *and* chain to reduce the number of layers.

File: .gitignore [hunk 11]
```diff
@@ -19,3 +19,6 @@ files/*
 .sass-cache
 static/css/*
 !static/css/.keep
+
+# Exclude flask config file as it contains secret keys
+instance/config.py
```
Comment (SantiagoTorres) [inline .gitignore]: Is this relevant to the docker image?

File: docker-start.sh [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+mongod &
+mkdir instance
+echo 'DEBUG=False' >> instance/config.py
```
Comment (SantiagoTorres) [inline docker-start.sh]: If this is meant for development, why do we need to set `DEBUG` to false?

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 10 above)
Comment (SantiagoTorres) [reply Dockerfile]: I also don't know if we want to use **four** different types of package managers, as they overlap their package offers and sometimes have different versions supported. I suggest that you:

1. use `apt's` [python-flask modules where possible.](https://packages.ubuntu.com/search?keywords=python-flask) Specially with such a simple requirements file.
2. use apt's gulp/other dependencies, and/or move the gulp part before it.
3. Use apt's bindings for libsass in python or whatever (which I assume is happening here). You are installing **three** more package managers for no reason at all (and again, it will result in a way bigger image). 

This whole thing also makes me wonder about the design of the whole stack, we can probably limit this to one/two langauges :P

File: Dockerfile (hunk 4 above)
Comment (ShikherVerma) [reply Dockerfile]: `gem` comes with `ruby` and `ruby-dev` was needed to install `sass` gem. Maybe I could ditch this in favour of some `apt` package.  
`nodejs` is for [npm dependencies](https://github.com/in-toto/layout-web-tool/blob/wizard-beta/gulpfile.js) not sure if they would be available on `apt` but I'll look.

File: Dockerfile (hunk 3 above)
Comment (ShikherVerma) [reply Dockerfile]: So chaining them with `&&`

File: Dockerfile (hunk 1 above)
Comment (ShikherVerma) [reply Dockerfile]: nope. Will switch to `debian:stable-slim`.

File: Dockerfile (hunk 6 above)
Comment (ShikherVerma) [reply Dockerfile]: Not sure what `hkps` is. Will look into it.

File: docker-start.sh (hunk 12 above)
Comment (ShikherVerma) [reply docker-start.sh]: Not sure what this `DEBUG` does. @lukpueh should I make it `True` ?

File: docker-start.sh (hunk 8 above)
Comment (ShikherVerma) [reply docker-start.sh]: I could generate a random key every time this script is run ?

File: .gitignore (hunk 11 above)
Comment (ShikherVerma) [reply .gitignore]: nope. not really. but I think its useful to have this here so people don't commit there secret key by mistake ?

File: Dockerfile (hunk 10 above)
Comment (lukpueh) [reply Dockerfile]: Re slimmer stack: Sure, what do you have in mind? I guess we can scrap Ruby and use a different sass implementation...
 
We'll probably have to keep Python (for back-end) and node (for front-end dependency management) unless you know a good Python-based front-end management system, or want to port the back-end to node? :P

File: docker-start.sh (hunk 12 above)
Comment (lukpueh) [reply docker-start.sh]: It's a toggle to enable/disable [`DEBUG` mode](http://flask.pocoo.org/docs/0.12/quickstart/#debug-mode), which shows additional debugging information and should be disabled in production. 
I have set it to [`TRUE` in the app](https://github.com/in-toto/layout-web-tool/blob/wizard-beta/wizard.py#L76) and override it with `FALSE` in the production [config file](http://flask.pocoo.org/docs/0.12/config/#configuring-from-files).

Comment (lukpueh) [reply docker-start.sh]: Is this dockerfile meant for development? I guess in *local/private deployment* the leaking sensitive info part is not really a problem.

Comment (lukpueh) [reply docker-start.sh]: Right now, the default settings (hardcoded in `wizard.py`) don't make a lot of sense because the app fails if no `instance/config.py` is provided. 
My reasoning for letting the app fail was that I didn't want anyone to use the default settings in production without noticing.

File: docker-start.sh (hunk 8 above)
Comment (lukpueh) [reply docker-start.sh]: Yes, hardcoded production secret key - not so good. Randomly creating a new one on every start isn't probably a good idea either. The key is used to sign user sessions, so in case of a restart users would lose their ongoing sessions. Although this might be a corner case, I think we can find a better solution. Just search the web for how others create Flask app docker files.

File: Dockerfile (hunk 10 above)
Comment (lukpueh) [reply Dockerfile]: Basically `node/npm/gulp` and `ruby/sass` are only needed whenever the front-end dependencies or the `scss` code changes. 

During runtime we only need `Python/Flask/mod_wsgi` and `mongodb`.  A multi-stage build could separate these things.

Also note, that your dockerfile actually misses the front-end management calls, i.e.  a run of the `gulp`-script and the `scss` to `css` compilation.

File: .gitignore (hunk 11 above)
Comment (SantiagoTorres) [reply .gitignore]: I also wonder if config.py will hold more variables in the future and it'll be a mess. I think we're better off by not adding this change to the repo.

File: Dockerfile (hunk 3 above)
Comment (SantiagoTorres) [reply Dockerfile]: yep

File: docker-start.sh (hunk 12 above)
Comment (SantiagoTorres) [reply docker-start.sh]: We probably want to revamp this whole image later to host the mongo daemon/db in a separate container. I'd consider this more of a helpful way to deploy for local testing...

File: docker-start.sh (hunk 9 above)
Comment (ShikherVerma) [reply docker-start.sh]: It won't matter as `mkdir` is not going to cost any significant time.

File: docker-start.sh (hunk 12 above)
Comment (ShikherVerma) [reply docker-start.sh]: leaving `DEBUG=False` here. Assuming Dockerfile just for local deployment and testing.

File: docker-start.sh (hunk 8 above)
Comment (ShikherVerma) [reply docker-start.sh]: Couldn't find anything regarding secret keys for dockerized flask app.

**Comment by ShikherVerma** [comment]
Pull image using [`docker pull shikherverma42/in-toto-layout-tool`](https://hub.docker.com/r/shikherverma42/in-toto-layout-tool/tags/)
size `239 MB`

File: docker-start.sh (hunk 9 above)
Comment (SantiagoTorres) [reply docker-start.sh]: `O(1)` vs `O(n)` some would say...

**Comment by ShikherVerma** [comment]
This pull request is made to the `beta` branch.
Since that branch has been merged to the `develop` branch should I close this and make a new PR to the `develop` branch ?

**Comment by SantiagoTorres** [comment]
@ShikherVerma FYI: we just changed the base branch to develop.

**Comment by ShikherVerma** [comment]
Oh I never used this feature of github. Its been so many years. How could I miss this :astonished:

**Comment by lukpueh** [comment]
closing due to inactivity
